File: day22.py
from collections import deque

seeds = [int(line.rstrip()) for line in open('test.in')]

def pseudo_random(n):
    n = ((n << 6) ^ n) & ((1 << 24) - 1)
    n = ((n >> 5) ^ n) & ((1 << 24) - 1)
    n = ((n << 11) ^ n) & ((1 << 24) - 1)
    return n

res = 0
by_deltas = {}
for seed in seeds:
    seen = set()
    n = seed
    prev = n % 10
    deltas = deque()
    for _ in range(2000):
        n = pseudo_random(n)
        price = n % 10
        deltas.append(price - prev)
        if len(deltas) > 4:
            deltas.popleft()
        if len(deltas) == 4:
            key = (deltas[0], deltas[1], deltas[2], deltas[3])
            if key not in seen:
                seen.add(key)
                if key not in by_deltas:
                    by_deltas[key] = price
                else:
                    by_deltas[key] += price
        prev = price
    res += n

print(res)
print(max(by_deltas.values()))

File: day22.rs
use std::{collections::{VecDeque, HashMap, HashSet}, io::{self, BufRead}, fs::File};

fn pseudo_random(mut n: u32) -> u32 {
    n = ((n << 6) ^ n) & ((1 << 24) - 1);
    n = ((n >> 5) ^ n) & ((1 << 24) - 1);
    ((n << 11) ^ n) & ((1 << 24) - 1)
}

fn main() {
    let f = File::open("test.in").expect("file not found");
    let mut res = 0u64;
    let mut seeds = Vec::new();
    let mut by_deltas: HashMap<_, i32> = HashMap::new();
    for line in io::BufReader::new(f).lines() {
        let seed = line.expect("error").trim().parse::<u32>().unwrap();
        seeds.push(seed);
        let mut n = seed;
        let mut deltas = VecDeque::new();
        let mut seen = HashSet::new();
        let mut prev = (n % 10) as i32;
        for _ in 0..2000 {
            n = pseudo_random(n);
            let price = (n % 10) as i32;
            deltas.push_back(price - prev);
            if deltas.len() > 4 {
                deltas.pop_front();
            }
            if deltas.len() == 4 {
                let key = (deltas[0], deltas[1], deltas[2], deltas[3]);
                if !seen.contains(&key) {
                    seen.insert(key);
                    *by_deltas.entry(key).or_default() += price;
                }
            }
            prev = price;
        }
        res += n as u64;
    }

    println!("{}", res);
    println!("{}", by_deltas.values().max().unwrap());
}
